### PR TITLE
feat: Add ability to echo request headers in trailing metadata

### LINKED ISSUE #1493: Add support to echo back the `x-goog-api-version` header so that AIP-4236 can be tested
See https://github.com/aip-dev/google.aip.dev/pull/1331 for AIP 4236 and [review feedback](https://github.com/googleapis/gapic-showcase/pull/1484#issuecomment-2043734889) from #1484

> the showcase server "echos" back the metadata/headers it was sent as trailing metadata/response headers so that the generator integration tests can capture them and assert what it sent was expected

### PR DESCRIPTION
**Pull request by parthea**
~This includes fixes from https://github.com/googleapis/gapic-showcase/pull/1500. Please review https://github.com/googleapis/gapic-showcase/pull/1500 first~ Done

Fixes #1493

Using this showcase server, I was able to see the http request headers in the trailing metadata

Response
```
responses.trailing_metadata()[0]
_Metadatum(key='x-goog-api-version', value='v1_20240408')

```

### CODE REVIEW

File: schema/google/showcase/v1beta1/echo.proto [hunk 1]
```diff
@@ -222,6 +222,9 @@ message EchoRequest {
   optional string other_request_id = 8 [
     (google.api.field_info).format = UUID4
   ];
+
+  // Http request headers to echo via EchoResponse.http_request_header_value
```
Comment (noahdietz) [inline schema/google/showcase/v1beta1/echo.proto:226]: I am not sure we need a field for this, I think sending everything that appears in the request back is fine.

File: schema/google/showcase/v1beta1/echo.proto [hunk 2]
```diff
@@ -237,6 +240,13 @@ message EchoResponse {
 
   // The other request ID specified or autopopulated in the request.
   string other_request_id = 4;
+
+
+  message RepeatedValues {
+    repeated string header_values = 1;
+  }
+  // The HTTP header to echo as specified in the request
+  map<string, RepeatedValues> http_request_header_value = 5;
```
Comment (noahdietz) [inline schema/google/showcase/v1beta1/echo.proto:249]: Why not just echo it back in the response headers directly? That's what we've done in the past...

File: server/services/echo_service.go [hunk 3]
```diff
@@ -46,12 +46,21 @@ type echoServerImpl struct {
 
 func (s *echoServerImpl) Echo(ctx context.Context, in *pb.EchoRequest) (*pb.EchoResponse, error) {
 	err := status.ErrorProto(in.GetError())
-	if err != nil {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if err != nil || !ok {
 		return nil, err
 	}
```
[the remark this answers is not in the document]
Comment (parthea) [reply server/services/echo_service.go]: Addressed in a37310e5304cfac0f67c6a13c7ce5b4512ecb30a

[the remark this answers is not in the document]
Comment (parthea) [reply server/services/echo_service.go]: Addressed in https://github.com/googleapis/gapic-showcase/pull/1501/commits/200897b4e45b3982544d6480f93ddb9a3ac1c827

File: schema/google/showcase/v1beta1/echo.proto (hunk 1 above)
Comment (parthea) [reply schema/google/showcase/v1beta1/echo.proto]: Fixed in https://github.com/googleapis/gapic-showcase/pull/1501/commits/51151734e80fb6eb15a7d88c09a8c7a080e53ff0

Updated response based on local testing:
```
response
content: "The hail in Wales falls mainly on the snails."
request_id: "some_value"
http_request_header_value {
  key: "x-goog-api-version"
  value {
    header_values: "v1_20240408"
  }
}
http_request_header_value {
  key: "x-goog-api-client"
  value {
    header_values: "gl-python/3.9.16 grpc/1.63.0.dev0 gax/2.18.0 gapic/0.0.0"
  }
}
http_request_header_value {
  key: "user-agent"
  value {
    header_values: "grpc-python/1.63.0.dev0 grpc-c/39.0.0 (linux; chttp2)"
  }
}
http_request_header_value {
  key: "grpc-accept-encoding"
  value {
    header_values: "identity, deflate, gzip"
  }
}
http_request_header_value {
  key: "content-type"
  value {
    header_values: "application/grpc"
  }
}
http_request_header_value {
  key: ":authority"
  value {
    header_values: "localhost:7469"
  }
}
```

File: schema/google/showcase/v1beta1/echo.proto (hunk 2 above)
Comment (parthea) [reply schema/google/showcase/v1beta1/echo.proto]: Fixed in https://github.com/googleapis/gapic-showcase/pull/1501/commits/8d1f1031e539275c0be3e0812867e7fcb75a81eb.

See trailing metadata below
```
> responses.trailing_metadata()[0]
_Metadatum(key='x-goog-api-version', value='v1_20240408')
```

**Comment by parthea** [comment]
This works for gRPC. I haven't figured out how to get the information for REST

**Comment by parthea** [comment]
@noahdietz Please could you take another look? Should we also support this for REST?

File: server/genproto/compliance.pb.go
Comment (noahdietz) [inline server/genproto/compliance.pb.go:1]: Please revert all of the changes in the `server/genproto` directory, the headers will just be replaced when things are regenerated by CI.

**Comment by noahdietz** [comment]
> Should we also support this for REST? 

Yes, definitely.

> Can that be added in a separate PR?

Yes, definitely.

We may want the generated REST handlers, like this: https://github.com/googleapis/gapic-showcase/blob/a34fc375c1e3c1ef0b6d79323c969cef045f9de4/server/genrest/echo.go#L87-L99

...to pull the same metadata from the context (as the gRPC handlers do) and write them as HTTP headers. I'd probably add a helper [here](https://github.com/googleapis/gapic-showcase/blob/main/util/genrest/resttools/headers.go) for copying context metadata into HTTP response headers, then invoke that in the spot above (generator code [here](https://github.com/googleapis/gapic-showcase/blob/a34fc375c1e3c1ef0b6d79323c969cef045f9de4/util/genrest/goviewcreator.go#L226-L244)).

@vchudnov-g may be interested in this.

File: server/genproto/compliance.pb.go
Comment (parthea) [reply server/genproto/compliance.pb.go:1]: Done

**Review by noahdietz — APPROVED** [review]

**Pull request merged**